staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -296,7 +296,7 @@ func (wc *watchChan) startWatching(watchClosedCh chan struct{}) {
 	if wc.recursive {
 		opts = append(opts, clientv3.WithPrefix())
 	}
-	if wc.progressNotify {
+	if wc.internalPred.AllowWatchBookmarks {
```

Comment:
Maybe I'm a bit paranoid, but I'm slightly worried about this change.

Up until now we have AllowWatchBookmarks and ProgressNotify separated.
This meant that:
- user could request bookmarks
- but if watchcache was disabled and that was going directly to etcd3 layer, it was ignored (and progress notify events were not returned back to the user).

With this change, the above behavior changes.
With watchcache, we send bookmarks ~every 1m, but progress notify are configured to lower values at least in some environments (like 5s).
This means, that if someone has watchcache disabled, they will suddenly may start getting 10x+ bookmarks. (which is obviously overhead on control-plane too).

At the very least, this will be the change for Events, for which watchcache is by default disabled.

I would like to see a convincing rationale why this is ok...

Reply:
> but if watchcache was disabled and that was going directly to etcd3 layer, it was ignored (and progress notify events were not returned back to the user).

yes, I thought this is a bug :) I didn't know it was intentional.

Today when you do `k get --raw '/api/v1/namespaces/default/events?watch=true&allowWatchBookmarks=true'` against a cluster built from the `master` branch you won't get any bookmark events.

Reply:
It's not necessary a bug. From this KEP:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/956-watch-bookmark#proposal

"
We consciously make it just a boolean flag - this gives kube-apiserver an ability to choose when they should be send without setting any expectations on user side how frequently and when it would be happening. In particular client isn't guaranteed to get any bookmarks. 
"

Reply:
ok, do you see value in adding a mechanism similar to the one we have in the cache for bookmarks? 

I think it would be nice to have feature parity between different implementations.

Reply:
yes, we already have different semantics depending on the caller:
- if watchcache then same-as-above
- if an api client then don't send any bookmark events

in addition to that we are about to add one more variation to the above list:
- if an api client requesting streaming then send only one special bookmark event


Let me bring back the progressNotify field add add some documentation if that is okay with you.

Reply:
pushed, PTAL.

Reply:
I have also opened https://github.com/kubernetes/kubernetes/issues/120348